LSPS0/README.md

```diff
@@ -0,0 +1,603 @@
+  client and one LSP, which are developed by at least two different
+  organizations or open-source project teams, and both development
+  teams have reported interoperability without further modifications
+  or clarifications of the specification.
+
+## Motivation
+
+The transport layer describes how a client is able to request
+services from the LSP.
+
+### Actors
+
+'LSP' is the API provider, and acts as the server.
+'Client' is the API consumer.
+
+## Protocol
+
+### Lightning Peer-To-Peer Transport
+
+The Lightning Network [BOLT8][] specification describes the transport
+layer for Lightning Network peer-to-peer messages.
+
+[BOLT8]: https://github.com/lightning/bolts/blob/f7dcc32694b8cd4f3a1768b904f58cb177168f29/08-transport.md
+
+Lightning Network BOLT8 describes messages as having two components:
+
+* A 16-bit message ID (encoded in big-endian).
+* A variable-length message payload (the remainder of the message).
+
+LSPS messages and access to LSP endpoints is via BOLT8 messages.
+
+> **Rationale** All clients and LSPs are expected to be Lightning
+> Network participants, and need to somehow talk using the BOLT8
+> protocol.
+> Thus, this is not an additional dependency for any client and
+> any LSP, the way an HTTP(S) or gRPC protocol would pull in
+> additional dependencies.
+>
+> During development of this specification, onion messages were
+> proposed.
+> As a client needs to connect to the LSP node and manage channels
+> with that node anyway, and LSP nodes want to be easily contactable
+> from IPv4, IPv6, DNS, or TorV3 contact points, the ability of
+> onion messages to send to a remote Lightning Network node (that
+> might not be directly contactable) was deemed unnecessary for
+> most client-LSP communication needs.
+
+For LSPS0, the message ID SHALL be 37913 (`lsps0_message_id`).
+All LSPS messages use only this message ID.
+
+> **Rationale** We indicate a single message ID as this reduces
+> the "footprint" of all LSPS specifications to only that single
+> message ID, increasing the probability that other protocols
+> using the Lightning BOLT8 peer-to-peer transport will not
+> conflict with the LSPS specifications.
+> The message ID 37913 is odd in order to comply with the [it's
+> OK to be odd][] rule, and is in the 32768 to 65535 range reserved
+> for extensions of the BOLT protocol.
+> Some implementations, such as Core Lightning, only expose
+> odd-numbered messages to custom message handlers, while others,
+> such as LDK, only expose 32768 and above.
+> The message ID was otherwise selected by random and has no
+> special meaning.
+
+[it's OK to be odd]: https://github.com/lightning/bolts/blob/master/00-introduction.md#its-ok-to-be-odd
+
+The payload contains the UTF-8 encoding of a complete JSON
+object.
+
+If a client or LSP receives a BOLT8 message with message ID
+37913:
+
+* If the payload does not parse as a complete JSON object
+  (i.e. a JSON key-value store or dictionary), MUST fail
```

Comment:
The "MUST fail" is only for LSPs, as it is the LSPs who send responses, and you can only signal a failure via a response.  If the LSP sends some really weird data to the client, the only thing the client can really do is to not use the LSP as. an LSP.

Reply:
Maybe I should re-word this entire section to something like this:

```Markdown
Both clients and LSPs MUST perform the following checks.
If any of these checks fail, then the message has failed with a
"bad message format" error.

* The payload MUST parse as a complete JSON object (i.e.
  a JSON key-value store or dictionary).
* The payload MUST NOT parse as more than a single
  complete JSON object, plus optional whitespace.
* The payload MUST NOT have any 0 bytes.
* If the client, the JSON object MUST be a JSON-RPC 2.0
  response or notification object.
* If the LSP, the JSON object MUST be a JSON-RPC 2.0
  request object.

In case of a "bad message format" error above, the client:

* MUST ignore the message.
* SHOULD log this as an unusual event.
* SHOULD avoid using the peer as an LSP.

In case of a "bad message format" error above, the LSP:

* MUST respond with a parse error (`code` = `-32602`,
  `id` = `null`) response object.
* MUST ignore the message.
* SHOULD log this as an unusual event.
```